RUFAS/biophysical/animal/pen.py

```diff
@@ -846,25 +852,117 @@ def formulate_optimized_ration(
         None
 
         """
+        info_map = {
+            "class": "Pen",
+            "function": self.formulate_optimized_ration.__name__,
+        }
+        if self.animal_combination == AnimalCombination.LAC_COW:
+            self.reset_milk_production_reduction()
 
-        # optimized_ration = optimize_ration(
-        #     available_feeds,
-        #     self.average_animal_requirements,
-        #     max_daily_feeds,
-        #     advance_purchase_allowance,
-        #     total_inventory
-        # )
-        optimized_ration: dict[RUFAS_ID, float] = {}  # Maps RuFaS Feed ID to mass of feed in ration per animal per day.
+        self.set_animal_nutritional_requirements(temperature=temperature, available_feeds=pen_available_feeds)
+        previous_ration = None
+        if hasattr(self, "ration"):
+            previous_ration = self.ration
+        solution, ration_config = ration_optimizer.attempt_optimization(
+            pen_average_body_weight=self.average_body_weight,
+            requirements=self.average_nutrition_requirements,
+            pen_available_feeds=pen_available_feeds,
+            animal_combination=self.animal_combination,
+            previous_ration=previous_ration,
+        )
+        num_attempts: int = 1
+        if solution and not solution.success:
+            ration_optimizer.handle_failed_constraints(
+                num_attempts=num_attempts,
+                solution=solution,
+                ration_config=ration_config,
+                animal_combination=self.animal_combination,
+                pen_id=self.id,
+                pen_available_feeds=pen_available_feeds,
+                average_nutrient_requirements=self.average_nutrition_requirements,
+                sim_day=9999,
+                info_map=info_map,
+            )
+            # TODO get sim day from RuFaS time
+        if self.animal_combination == AnimalCombination.LAC_COW:
+            while not solution.success:
+                if self.average_milk_production < AnimalModuleConstants.MINIMUM_AVG_PEN_MILK:
+                    om.add_error(
+                        "Milk production too low",
+                        (
+                            f"Check failed_constraint_summary_for_pen_{self.id} to see what caused formulation to fail."
+                            f" Possible solution is to provide additional feed ingredients to "
+                            f"{self.animal_combination.name}."
+                        ),
+                        info_map,
+                    )
+                    raise ValueError
```

Comment:
```suggestion
                    raise ValueError(                            f"Check failed_constraint_summary_for_pen_{self.id} to see what caused formulation to fail."
                            f" Possible solution is to provide additional feed ingredients to "
                            f"{self.animal_combination.name}.")
```
Or some abbreviated version of the error message just to have something informative in the raised error. Same with the next ValueError